presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java

```diff
@@ -0,0 +1,252 @@
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.concat;
+
+public class WindowInfo
+        implements Mergeable<WindowInfo>, OperatorInfo
+{
+    private final List<DriverWindowInfo> windowInfos;
+
+    @JsonCreator
+    public WindowInfo(@JsonProperty("windowInfos") List<DriverWindowInfo> windowInfos)
+    {
+        this.windowInfos = ImmutableList.copyOf(windowInfos);
+    }
+
+    @JsonProperty
+    public List<DriverWindowInfo> getWindowInfos()
+    {
+        return windowInfos;
+    }
+
+    @Override
+    public WindowInfo mergeWith(WindowInfo other)
+    {
+        return new WindowInfo(ImmutableList.copyOf(concat(this.windowInfos, other.windowInfos)));
+    }
+
+    @ThreadSafe
+    static class DriverWindowInfoBuilder
+    {
+        private ImmutableList.Builder<IndexInfo> indexInfosBuilder = ImmutableList.builder();
+        private IndexInfoBuilder currentIndexInfoBuilder = null;
+
+        public synchronized void addIndex(PagesIndex index)
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
+                if (indexInfo.isPresent()) {
+                    indexInfosBuilder.add(indexInfo.get());
+                }
+            }
+            currentIndexInfoBuilder = new IndexInfoBuilder(index.getPositionCount(), index.getEstimatedSize().toBytes());
+        }
+
+        public synchronized void addPartition(WindowPartition partition)
+        {
+            checkState(currentIndexInfoBuilder != null, "addIndex must be called before addPartition");
+            currentIndexInfoBuilder.addPartition(partition);
+        }
+
+        public synchronized DriverWindowInfo build()
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
+                if (indexInfo.isPresent()) {
+                    indexInfosBuilder.add(indexInfo.get());
+                }
+                currentIndexInfoBuilder = null;
+            }
+
+            List<IndexInfo> indexInfos = indexInfosBuilder.build();
+            if (indexInfos.size() == 0) {
+                return new DriverWindowInfo(0.0, 0.0, 0.0, 0, 0, 0);
+            }
+            long totalRowsCount = indexInfos.stream()
+                    .mapToLong(index -> index.getTotalRowsCount())
+                    .sum();
+            double averageIndexPositions = totalRowsCount / indexInfos.size();
+            double squaredDifferencesPositionsOfIndex = indexInfos.stream()
+                    .mapToDouble(index -> Math.pow(index.getTotalRowsCount() - averageIndexPositions, 2))
+                    .sum();
+            double averageIndexSize = indexInfos.stream()
+                    .mapToLong(index -> index.getSizeInBytes())
+                    .average().orElse(Double.NaN);
```

Comment:
Then use `getAsDouble` because it will throw an exception if the value is not present (which is unexpected).